autosklearn/data/target_validator.py

```diff
@@ -32,68 +59,91 @@ def __init__(self) -> None:
+                                     np.shape(y_test)
+                                 ))
+            if isinstance(y_train, pd.DataFrame):
+                y_train = typing.cast(pd.DataFrame, y_train)
+                y_test = typing.cast(pd.DataFrame, y_test)
+                if y_train.columns != y_test.columns:
+                    raise ValueError(
+                        "Train and test targets must both have the same columns, yet "
+                        "y={} and y_test={} ".format(
+                            y_train.columns,
+                            y_test.columns
+                        )
+                    )
+
+                if list(y_train.dtypes) != list(y_test.dtypes):
+                    raise ValueError("Train and test targets must both have the same dtypes")
+
+        if self.out_dimensionality is None:
+            self.out_dimensionality = 1 if len(shape) == 1 else shape[1]
+        else:
+            _n_outputs = 1 if len(shape) == 1 else shape[1]
+            if self.out_dimensionality != _n_outputs:
+                raise ValueError('Number of outputs changed from %d to %d!' %
+                                 (self.out_dimensionality, _n_outputs))
 
         # Fit on the training data
-        self._fit(y_train, y_test, is_classification)
+        self._fit(y_train, y_test)
 
         self._is_fitted = True
 
         return self
 
     def _fit(
         self,
-        y_train: typing.Union[np.ndarray, pd.DataFrame],
-        y_test: typing.Optional[typing.Union[np.ndarray, pd.DataFrame]] = None,
-        is_classification: bool = False,
+        y_train: SUPPORTED_TARGET_TYPES,
+        y_test: typing.Optional[SUPPORTED_TARGET_TYPES] = None,
     ) -> BaseEstimator:
         """
         If dealing with classification, this utility encodes the targets.
 
-        It does so by also using the classes from the test data, to prevent enconding
+        It does so by also using the classes from the test data, to prevent encoding
         errors
 
         Parameters
         ----------
-            y_train: typing.Union[np.ndarray, pd.DataFrame]
+            y_train: SUPPORTED_TARGET_TYPES
                 The labels of the current task. They are going to be encoded in case
                 of classification
-            y_test: typing.Union[np.ndarray, pd.DataFrame]
+            y_test: typing.Optional[SUPPORTED_TARGET_TYPES]
                 A holdout set of labels
-            is_classification: bool
-                A bool that indicates if the validator should operate in classification mode.
-                During classification, the targets are encoded.
         """
-
-        if not hasattr(y_train, 'shape'):
-            # Make it numpy array for easier checking
-            y_train = np.array(y_train)
-
-        if not is_classification or self.type_of_target == 'multilabel-indicator':
+        if not self.is_classification or self.type_of_target == 'multilabel-indicator':
```

Comment:
Would it be possible that you add checks for these different cases which are not tested according to codecov?